index.html

```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        </p>
+
+        <p>
+        Operations that happens <dfn data-lt="atomic">atomically</dfn> on a given thread, can only
+        be executed when no other <a>atomic</a> operation is
+        running on another thread.
+        </p>
+
+        <p>
+        The algorithm for rendering a block of audio from an <a>AudioContext</a>
+        <em>G</em> with an <a href=#control-message-queue>control message queue</a>
+        <em>Q</em> is as follow:
+        </p>
+        <ol>
+          <li> Let <em>Q<sub>rendering</sub></em> be an empty <a
+            href=#control-message-queue>control message queue</a>.
+          <a>Atomically</a> swap <em>Q<sub>rendering</sub></em> and
+          <em>Q</em></li>
+        <li> Until there is no more event in <em>Q<sub>rendering</sub></em>,
+        execute the following steps:
+        <ol>
+          <li>
+          Execute the asynchronous section of the oldest element of
+          <em>Q<sub>rendering</sub></em>.
+          </li>
+          <li>
+          Remove the oldest element of <em>Q<sub>rendering</sub></em>.
+          <li>Let <em>Q<sub>rendering</sub></em> be an empty <a>control message
+          queue</a>. <a>Atomically</a> <a>swap</a> <em>Q<sub>rendering</sub></em> and
+          <em>Q</em>
+          </li>
+          <li>Until there is no more messages in <em>Q<sub>rendering</sub></em>,
+          execute the following steps:
+            <ol>
+              <li>Execute the asynchronous section of the <a>oldest message</a>
+              of <em>Q<sub>rendering</sub></em>.
+              </li>
+              <li>Remove the <a>oldest message</a> of
+              <em>Q<sub>rendering</sub></em>.  </li>
+            </ol>
+          </li>
+          <li>Order the <a>AudioNode</a>s of the <a>AudioContext</a> to be
+          processed.
+            <ol>
+              <li>Let <em>ordered</em> be an empty list of <a>AudioNode</a>s.
+              It will contain an ordered list of <a>AudioNode</a>s when this
+              ordering algorithm terminates.
+              </li>
+              <li>Let <em>nodes</em> be the set of all nodes belonging to an
+              <a>AudioContext</a>.
+              </li>
+              <li>Let <em>cycle breakers</em> be an empty set of
+              <a>DelayNode</a>s. It will contain all the <a>DelayNode</a>s that
+              are part of a cycle.
+              </li>
+              <li>For each <a>AudioNode</a> <em>node</em> of <em>nodes</em>
+                <ol>
+                  <li>
+                    <em>node</em> is a <a>DelayNode</a> part of a cycle, add it
+                    to <em>cycle breakers</em> and remove it from
+                    <em>nodes</em>.
+                  </li>
+                </ol>
+              </li>
+              <li>If <em>nodes</em> contains cycles, mute all the
+              <a>AudioNode</a>s that are part of this cycle, and remove them
+              from <em>nodes</em>.
```

Comment:
It really feels as though cycle detection could occur in conjunction with the node-visiting-and-marking scheme by maintaining a list of nodes already encountered during a given recursive node traversal; if any of these get visited, then a cycle has been found and that list will hold the contents of the cycle.